tests/test_portfolio_env.py

```diff
@@ -0,0 +1,275 @@
+                                                    [[1, 2], [1e-9, 1e-6]],
+                                                    [[3, 4], [1, 2]]])
+
+
+def test_calculate_returns_as_observations_when_configured(make_env, datums):
+    datums.add().rows(
+        [2, 1, 4],
+        [4, 2, 8],
+        [1, 4, 2],
+        [1, 1, 1],
+    )
+    assert_obs_eq(make_env(window_size=2, calc_returns=True).reset(), [[2, 0.25],
+                                                                       [2, 2],
+                                                                       [2, 0.25]])
+
+
+def test_stepping_the_environment_returns_next_observation(make_ready_env, datums):
+    datums.add().rows([1], [2], [3])
+    env = make_ready_env()
+    assert_obs_eq(unpack_obs(idle_step(env)), [2])
+    assert_obs_eq(unpack_obs(idle_step(env)), [3])
+
+
+def idle_step(env):
+    a = normalize(env.portfolio.assets)
+    return env.step(a)
+
+
+def test_resetting_the_environment_resets_observations(make_ready_env, datums):
+    datums.add().rows([1], [2], [3])
+    env = make_ready_env()
+    assert_obs_eq(unpack_obs(idle_step(env)), [2])
+    env.reset()
+    assert_obs_eq(unpack_obs(idle_step(env)), [2])
+
+
+def test_environment_is_done_at_the_last_observation(make_ready_env, datums):
+    datums.add().rows([1], [2], [3])
+    env = make_ready_env()
+    assert not unpack_done(idle_step(env))
+    assert unpack_done(idle_step(env))
+
+
+def test_raise_an_error_when_not_enough_data_for_a_single_step(make_env, datums):
+    datums.add().rows([1], [2])
+    env = make_env(window_size=2)
+    with pytest.raises(DatumsError):
+        env.reset()
+
+
+def test_raise_an_error_when_stepping_past_done(make_ready_env, datums):
+    datums.add().rows([1], [2])
+    env = make_ready_env()
+    list(until_done(env, [1, 0]))
+    with pytest.raises(DatumsError):
+        idle_step(env)
+
+
+def test_stepping_with_returns(make_ready_env, datums):
+    datums.add().rows(
+        [2, 1, 4],
+        [4, 2, 8],
+        [1, 4, 2],
+        [1, 1, 1],
+    )
+    env = make_ready_env(window_size=2, calc_returns=True)
+    assert_obs_eq(unpack_obs(idle_step(env)), [[0.25, 1],
+                                               [2, 0.25],
+                                               [0.25, 0.5]])
+
+
+@pytest.mark.parametrize('invalid', [0, np.nan, np.inf, -1e-5])
+def test_raise_error_when_upcoming_invalid_datum_is_encountered(make_ready_env, datums, invalid):
+    datums.add().rows([2], [1], [invalid])
+    env = make_ready_env()
+    with pytest.raises(DatumsError):
+        idle_step(env)
+
+
+@pytest.mark.parametrize('num_assets', [1, 2])
+def test_action_shape_is_dependent_on_number_of_assets(make_ready_env, datums, num_assets):
+    fill_datums(datums, num_assets, 1)
+    env = make_ready_env()
+    assert env.action_space == make_box(high=1, low=-1, shape=(num_assets + 1,))
+
+
+def test_idle_immediate_reward(make_ready_env, datums):
```

Comment:
remove the 'immediate' as the different reward types are now handles by wrappers